### PR TITLE
Add 2FA module to Nexus.

### PR DESCRIPTION
**Pull request by jianba**

### CODE REVIEW

**Comment by Anduin2017** [comment]
You shouldn't open a pull request before your branch is ready to be merged.

If you want to compare the two branches, consider using some tools like gitlens.

**Comment by jianba** [comment]
> You shouldn't open a pull request before your branch is ready to be merged.
> 
> If you want to compare the two branches, consider using some tools like gitlens.

Got it.

**Comment by jianba** [comment]
I had tried to marge the latest code but when I  add marge, my function will work. I checked the marge codes and it seems that it is ok.

**Comment by Anduin2017** [comment]
Is it currently working and available to review now?

File: Gateway/Services/UserAppAuthManager.cs [hunk 1]
```diff
@@ -44,6 +44,29 @@ public async Task<IActionResult> FinishAuth(GatewayUser user, FinishAuthInfo mod
             }
         }
 
+        public async Task<IActionResult> TwoFAFinishAuth(GatewayUser user, FinishAuthInfo model, bool forceGrant)
```
Comment (Anduin2017) [inline Gateway/Services/UserAppAuthManager.cs]: You should NOT copy the finish auth method. Copy and paste are not what a developer should do. Neither your view.

Comment (jianba) [reply Gateway/Services/UserAppAuthManager.cs]: > Is it currently working and available to review now?

Yes, the code now works. I recorded a video of the feature presentation. I tried to send it to you, but the school's network is a bit rubbish (occasionally broken), GG several times.

Comment (jianba) [reply Gateway/Services/UserAppAuthManager.cs]: > You should NOT copy the finish auth method. Copy and paste are not what a developer should do. Neither your view.

Got it, I'll fix it. By tomorrow afternoon。

**Comment by Anduin2017** [comment]
Please resolve all reviews so we can merge it.

File: Account/Views/Account/TwoFactorAuthentication.cshtml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                    @Localizer["TwoFactorAuthentication Settings"]
+                </div>
+                <ul class="list-group list-group-flush">
+                    <li class="list-group-item">
+                        @if (!Model.NewHas2FAKey)
+                        {
+                            <p>
+                                Click the button below, and follow the prompts to start to configure your Two Factor Authentication function.
+                            </p>
+                            <P>
```
Comment (Anduin2017) [inline Account/Views/Account/TwoFactorAuthentication.cshtml]: Use lower case!

File: Account/Views/Account/TwoFactorAuthentication.cshtml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                                <P>
+                                    You have enabled the 2fa function.
+                                    <br />You need to enter the Verify Code in the Authentication App when you log in.
+                                    <br />You can log in through recovery codes when you cannot use the Authentication App.
+                                    <br />Recovery Codes are unique. The same Recovery Codes can only be viewed once.
+                                    <br /> You can get your recovery codes in two ways
+                                    <br />1. When you have 2fa key and enter verify code, you will get your recovery codes.
+                                    <br />2. Click the button below to generate a new Verify Codes.
+                                </P>
+                                <P>
```
Comment (Anduin2017) [inline Account/Views/Account/TwoFactorAuthentication.cshtml]: Lower case p.

File: Account/Views/Account/VerifyTwoFACode.cshtml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        <li class="breadcrumb-item active">
+            @ViewData["Title"]
+        </li>
+    </ol>
+    <div class="col-md-8 pl-0">
+        <div class="card mb-3">
+            <div class="card-body">
+                <div class="card-header">
+                    <i class="fa fa-fw fa-address-card"></i>
+                    @Localizer["TwoFactorAuthentication Settings"]
```
Comment (Anduin2017) [inline Account/Views/Account/VerifyTwoFACode.cshtml]: Typo

File: Account/Views/Account/ViewTwoFAKey.cshtml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        <li class="breadcrumb-item active">
+            @ViewData["Title"]
+        </li>
+    </ol>
+    <div class="col-md-8 pl-0">
+        <div class="card mb-3">
+            <div class="card-body">
+                <div class="card-header">
+                    <i class="fa fa-fw fa-address-card"></i>
+                    @Localizer["TwoFactorAuthentication Settings"]
```
Comment (Anduin2017) [inline Account/Views/Account/ViewTwoFAKey.cshtml]: Typo

File: Gateway/Controllers/UserController.cs [hunk 6]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
             await _dbContext.SaveChangesAsync();
             return this.Protocol(ErrorType.Success, $"Successfully unbound your {model.ProviderName} account.");
         }
+
+        [HttpPost]
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Mark what this API gonna produce!

File: Gateway/Controllers/UserController.cs [hunk 7]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+        public async Task<JsonResult> ViewHas2FAkey(ViewHas2FAkeyAddressModel model)
+        {
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, t => t.ChangeBasicInfo);
+            bool ReturnValue = user.Has2FAKey;
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: lower case!

File: Gateway/Controllers/UserController.cs [hunk 8]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+            return Json(new AiurValue<bool>(ReturnValue)
+            {
+                Code = ErrorType.Success,
+                Message = "Successfully get the target user's Has2FAkey."
+            });
+        }
+
+        [HttpPost]
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Mark what this API gonna produce!

File: Gateway/Controllers/UserController.cs [hunk 9]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+        public async Task<JsonResult> ViewTwoFactorEnabled(ViewTwoFactorEnabledAddressModel model)
+        {
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, t => t.ChangeBasicInfo);
+            bool ReturnValue = user.TwoFactorEnabled;
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: lower case!

File: Gateway/Controllers/UserController.cs [hunk 10]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+            return Json(new AiurValue<bool>(ReturnValue)
+            {
+                Code = ErrorType.Success,
+                Message = "Successfully get the target user's TwoFactorEnabled."
+            });
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> View2FAKey(Get2FAKeyAddressModel model)
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Mark what this API gonna produce!

File: Gateway/Controllers/UserController.cs [hunk 11]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+        [HttpPost]
+        public async Task<IActionResult> SetTwoFAKey(Set2FAKeyAddressModel model)
+        {
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, t => t.ChangeBasicInfo);
+            if (false == user.Has2FAKey)
+            {
+                user.Has2FAKey = true;
+                await _userManager.UpdateAsync(user);
+            }
+            var ReturnValue = user.Has2FAKey;
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: lower case!

File: Gateway/Controllers/UserController.cs [hunk 12]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
...
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, t => t.ChangeBasicInfo);
+            var returnList = await _twoFAHelper.LoadSharedKeyAndQrCodeUriAsync(user, model);
+            return Json(new AiurCollection<Get2FAKeyAddressModel>(returnList)
+            {
+                Code = ErrorType.Success,
+                Message = "Successfully set the user's TwoFAKey!"
+            });
+        }
+
+        [HttpPost]
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Mark what this API gonna produce!

File: Gateway/Services/UserAppAuthManager.cs (hunk 1 above)
Comment (Anduin2017) [reply Gateway/Services/UserAppAuthManager.cs]: Please fix this! Do not copy-and-paste the finish auth method!

File: Gateway/Controllers/UserController.cs [hunk 13]
```diff
@@ -282,5 +293,155 @@ public async Task<IActionResult> UnBindSocialAccount(UnBindSocialAccountAddressM
             await _dbContext.SaveChangesAsync();
             return this.Protocol(ErrorType.Success, $"Successfully unbound your {model.ProviderName} account.");
         }
+
+        [HttpPost]
+        public async Task<JsonResult> ViewHas2FAkey(ViewHas2FAkeyAddressModel model)
+        {
+            var user = await _grantChecker.EnsureGranted(model.AccessToken, model.OpenId, t => t.ChangeBasicInfo);
```
Comment (Anduin2017) [inline Gateway/Controllers/UserController.cs]: Use a new permission to ensure the app can change two fa info! Not the `ChangeBasicInfo`

**Pull request merged**